imap_processing/lo/l2/lo_l2.py

```diff
@@ -587,6 +655,45 @@ def load_geometric_factor_data(species: str) -> pd.DataFrame:
     return lo_ancillary.read_ancillary_file(gf_file)
 
 
+def reduce_geometric_factor_dataset(species: str, esa_mode: int) -> xr.Dataset:
+    """
+    Get geometric factor data as xarray Dataset for a specific species and ESA mode.
+
+    This helper function loads geometric factor data, filters by ESA mode, converts
+    to xarray, and selects all 7 energy steps for vectorized operations.
+
+    Parameters
+    ----------
+    species : str
+        The species to load geometric factors for ("h" or "o").
+    esa_mode : int
+        ESA mode (0 for HiRes, 1 for HiThr).
+
+    Returns
+    -------
+    xarray.Dataset
+        Geometric factor data indexed by Observed_E-Step (1-7), containing all
+        columns from the geometric factor CSV file.
+    """
+    # Load geometric factor data for this species
+    gf_data = load_geometric_factor_data(species)
+
+    # Filter for the specific ESA mode
+    if "esa_mode" in gf_data.columns:
+        gf_data = gf_data[gf_data["esa_mode"] == esa_mode].copy()
+
+    # Convert to xarray Dataset indexed by energy step for vectorized selection
+    gf_ds = gf_data.set_index("Observed_E-Step").to_xarray()
+
+    # TODO: The ancillary data has repeated Observed_E-Step values with different
+    #    incident_E-Step values. Figure out how we get the correct row. For now,
+    #    just remove duplicate Observed_E-Step values.
```

Comment:
They said that we should only use `observed == incident` so (1, 1), (2, 2), etc. Does this do that or does it keep the lowest pair instead so (1, 1), (1, 2), (1, 3), etc.